client.go

```diff
@@ -279,3 +284,27 @@ func (c *client) updateBandwidthStats(url string, sentBytes int64,
 	updateStats(&urlStats, sentBytes, receivedBytes)
 	c.bandwidthStats[url] = urlStats
 }
+
+func (c *client) processExpiredAds(ticker *time.Ticker) {
+	for range ticker.C {
+		c.removeExpiredAds()
+	}
+}
+
+func (c *client) removeExpiredAds() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for adId, ad := range c.inProgressAds {
+		leaseExpirySecond, ok := ad["lease_expiry"]
+		if !ok {
+			continue
+		}
+
+		// We are dropping the expired ad here and not expiring,
+		// because ad server expires them automatically after 24hrs.
+		if int64(leaseExpirySecond.(float64)) < time.Now().Unix() {
+			delete(c.inProgressAds, adId)
```

Comment:
instead of calling `removeFromInProgressList` function, which acquires a lock and then deletes,
I'm acquiring the lock at the start of the function and delete the element with delete function.